File: 10-regular-expression-matching/10-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:        
        def dp(si, pi):
            if pi >= len(p): return si == len(s)
            if si >= len(s): return pi + 1 < len(p) and p[pi + 1] == '*' and dp(si, pi + 2)
            if (si, pi) not in memo:
                matched = p[pi] == '.' or p[pi] == s[si]
                if pi + 1 < len(p) and p[pi + 1] == '*':
                    memo[(si, pi)] = dp(si, pi + 2) or (matched and dp(si + 1, pi))
                else:
                    memo[(si, pi)] = matched and dp(si + 1, pi + 1)
            return memo[(si, pi)]
        
        memo = {}
        return dp(0, 0)

File: 10-regular-expression-matching/10-regular-expression-matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        vector<vector<int>> memo(s.size()+ 1, vector<int>(p.size() + 1));
        return dp(s, p, 0, 0, memo);
    }
    
    bool dp(const string& s, const string& p, int i, int j, vector<vector<int>>& memo) {
        if (j >= p.size()) return i == s.size();
        bool is_match = i < s.size() && (s[i] == p[j] || p[j] == '.');
        bool res;
        
        if (memo[i][j] != 0) return memo[i][j] == 1 ? true : false;
        
        if (j + 1 < p.size() && p[j + 1] == '*') res = dp(s, p, i, j + 2, memo) || (is_match && dp(s, p, i + 1, j, memo));
        else res = is_match && dp(s, p, i + 1, j + 1, memo);
        
        if (res) memo[i][j] = 1;
        else memo[i][j] = 2;
        
        return res;
    }
};